File: 887-minimum-cost-to-hire-k-workers/minimum-cost-to-hire-k-workers.py
class Solution(object):
    def mincostToHireWorkers(self, quality, wage, k):
        n = len(quality)
        x = sorted((((wage[i]+.0) /quality[i], quality[i]) for i in range(n)))
        h = []
        q = 0
        for i in range(k):
            heapq.heappush(h, -x[i][1])
            q += x[i][1]
        res = q * x[k-1][0]
        for i in range(k, n):
            q += x[i][1] + heapq.heapreplace(h, -x[i][1])
            res = min(res, x[i][0] * q)
        return res

File: 857-minimum-cost-to-hire-k-workers/minimum-cost-to-hire-k-workers.java
import java.util.*;

class Solution {
    public double mincostToHireWorkers(int[] quality, int[] wage, int k) {
        int n = quality.length;
        List<double[]> x = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            x.add(new double[]{((double) wage[i]) / quality[i], quality[i]});
        }
        Collections.sort(x, (a, b) -> Double.compare(a[0], b[0]));
        PriorityQueue<Double> h = new PriorityQueue<>();
        double q = 0;
        for (int i = 0; i < k; i++) {
            h.offer(-x.get(i)[1]);
            q += x.get(i)[1];
        }
        double res = q * x.get(k - 1)[0];
        for (int i = k; i < n; i++) {
            q += x.get(i)[1] - (-h.poll());
            h.offer(-x.get(i)[1]);
            res = Math.min(res, x.get(i)[0] * q);
        }
        return res;
    }
}

